README.md
[![Build Status](https://travis-ci.org/adamdsmith/geobird.png)](https://travis-ci.org/adamdsmith/geobird)

USFWS Disclaimer
================

This United States Fish & Wildlife Service (USFWS) code is provided on an "as is" basis and the user assumes responsibility for its use. USFWS has relinquished control of the information and no longer has responsibility to protect the integrity , confidentiality, or availability of the information. Any reference to specific commercial products, processes, or services by service mark, trademark, manufacturer, or otherwise, does not constitute or imply their endorsement, recomendation or favoring by USFWS. The USFWS seal and logo shall not be used in any manner to imply endorsement of any commercial product or activity by USFWS or the United States Government.

The `geobird` package
=====================

This package contains functions to read and parse eBird database requests (.txt), store them in a local SQLite database, and to query that database geographically. Users may optionally buffer the input geography (i.e., a `SpatialPolygonsDataFrame` object created from an ESRI shapefile) by any distance. Additionally, functions are available to visualize the monthly phenology of eBird records by species for a given geography and produce a spreadsheet with eBird records by season. In the latter, seasonal abundance codes are assigned to each species based on the proportion of complete eBird checklists (i.e., observers reported all species observed), provided some minimum number of checklists is present for a given geography and season. Current functionality is limited to the United States, but there is little to prevent the expansion of the methods to other areas or georeferenced species occurrence databases.

Installing geobird
------------------

The geobird package requires you to have [R](https://www.r-project.org/) (&gt;= 3.2.1) installed on your computer as well as [Rtools](https://cran.r-project.org/bin/windows/Rtools/). Both will require administrative priveleges but the installation of packages after this initial install will not.

With R and Rtools installed, install and load the MABM package to access its functionality:

    install.packages("devtools") # Only the first time
    devtools::install_github("adamdsmith/geobird") # Only the first time
    library(geobird)

Using geobird
-------------

Basic operations are illustrated below. For details on the arguments and options available in each function, see their associated help files.

### Parsing eBird queries from Cornell

We typically access historical eBird observations by requesting state-by-state queries of the eBird database hosted by Cornell. It comes as an often large tab-delimited text file (e.g., Florida's weighed in at nearly 4 GB and contained over 10 million records). We only need a subset of the 43 variables contained in the file and we also wish to exclude records (1) not identified to species, (2) duplicated on eBird checklists of multiple observers, and (3) non-approved (i.e., not yet vetted) records.

The `parse_ebird` function accomplishes these tasks and outputs a data structure amenable to local storage in a SQLite database. One need only direct the function to the appropriate tab-delimited text file.

Take Georgia, for example. Parsing leaves us with ~3.3 million (out of more than 4 million) statewide eBird records.

``` r
ga_birds <- parse_ebird("../Data/ebd_US-GA_relAug-2015.txt") # 1.5 GB; > 4M records; takes < 1 min
class(ga_birds)
[1] "data.table" "data.frame"
dim(ga_birds)
[1] 3270274      12
```

### Storing it locally for quick access

We store this output locally in a SQLite database. It's a more convenient format for permanent storage and we can query the entire southeastern eBird database geographically and quickly pass the result from the database to R. Storing the entire database in local memory is not practicable.

The `store_ebird` function does the trick, although this requires you to have already established a local SQLite database (see example in `help(store_ebird)`).

``` r
store_ebird(ga_birds)
```

As of version 0.3, you have the option of combining the parsing and storing into a single function call using the `file` argument in `store_ebird`:

``` r
store_ebird(file = "../Data/ebd_US-GA_relAug-2015.txt")
```

### Making a geographic query of the database

We currently have the parsed eBird records from all 10 Region 4 (southeast) states stored in the local SQLite database for making geographic queries. All we need are some polygons defining the areas we wish to query. In our case, this comes in the form of an ESRI shapefile of southeastern NWRs and fish hatcheries.

First, we load the shapefile into R:

``` r
SErefuges <- rgdal::readOGR("../GIS", "refuges", verbose = FALSE, stringsAsFactors = FALSE)
```

Now, we have some decisions to make. Do we want to query all refuges (or a subset) and do we want to extend the area around the boundary (i.e., create a buffer) to capture nearby observations not technically on the refuge?

Either way, the `geo_ebird` function is our friend.

The main arguments we need to worry with are `query_polys` (the polygons we want to query), `buffers` (the distance(s), in km, we wish to expand our polygon boundary; defaults to 0), and `which_polys` (the identification of specific polygons in `query_polys` if we don't want them all). For your particular application, you'll likely need to direct the function to the location of the local SQLite database with the `ebird_sqlite` argument and specify the column name in the shapefile that defines the polygons (i.e., the `poly_id` argument). See `?geo_ebird` for details on these arguments.

For example, we could query the eBird records in all southeastern refuges, with a 2 km buffer around each, using the following command:

``` r
all_nwrs <- geo_ebird(SErefuges, buffers = 2)
```

However, we'll take the less time-consuming example of using only a single refuge and no buffer (i.e., use the actual NWR boundary). In this case, we'll use Piedmont National Wildlife Refuge.

All we need to specify is the file containing the southeastern refuges polygons and select Piedmont NWR using the `which_polys` argument. This does require you to know what your polygons are called and to specify the name of the column containing that information, but because we built this around our southeastern NWR shapefile, we get to use the defaults. Lucky us! Furthermore, all we need is the name of the refuge; we can toss the formality of calling it a "National Wildlife Refuge" or "National Fish Hatchery."

``` r
piedmont <- geo_ebird(SErefuges, which_polys = "Piedmont")
```

This leaves us with just over 17,500 eBird records (species observations) within the actual boundary of Piedmont NWR.

We could easily have picked a few refuges as well. Notice the name is insensitive to capitalization. Here, for example, we choose to explore eBird records within the actual refuge boundaries and also with an area represented by the refuge and a 10 km buffer.

``` r
a_few_nwrs <- geo_ebird(SErefuges, which_polys = c("Piedmont", "HOBE SoUnD", "CEDar iSLand"),
                        buffer = c(0, 10))
```

### Checklists of species abundance and occurrence

The primary goal of this project was to output, for each refuge, a species list for each refuge and some consistent measure of relative abundance and occurrence.

The `make_checklists` function generates, for each queried polygon, a spreadsheet of all species detected on at least one eBird checklist. For each season (defined below), an estimated abundance classification (defined below) is provided if a minimum number of complete checklists is available (specified with the `min_lists` argument). Regardless, the total number of checklists (complete and incomplete) on which each species was observed in each season is reported. Likewise, the amount of effort (\# of complete and total checklists) is reported by season.

We classified seasons as follows: spring (Mar - May), summer (Jun - Aug), fall (Sep - Nov), and winter (Dec - Feb). This classification corresponds roughly to the seasonal bird checklists present in many NWR brochures (which may have had a small part in the decision to use this classification).

Abundance designations are classified based on the proportion of complete checklists on which a given species occurs. The `min_lists` argument allows the user to determine the number of checklists necessary to even attempt such a designation. The default is 10 complete checklists, although a larger minimum (e.g., 50 or 100) is probably a more reasonable choice for generating an estimate of species relative abundance. Abundance classifications (% of complete checklists) are: Abundant (&gt;= 20%), Common \[5 - 20%), Uncommon \[2.5 - 5%), Occasional \[1 - 2.5%), and Rare (&lt; 1%).

Continuing our Piedmont NWR example, the output of the `make_checklists` function is an Excel spreadsheet containing four sheets: (1) seasonal abundance codes for those season with enough complete checklists, (2) seasonal occurrence information giving the total number of checklists (complete and incomplete) reporting a given species, (3) a finalized checklist within the actual refuge boundary (if the actual boundary was requested) and (4) a summary of the eBird effort at each requested buffer distance (seasonal totals of complete and all eBird checklists).

If no seasons have enough complete checklists to generate abundance codes, the first and third sheets (seasonal abundance and species checklist) are not generated. With Piedmont NWR, this is irrelevant as all seasons contain a sufficient number of complete checklists (as we've defined it; 10 complete checklists).

The columns included on the seasonal abundance and occurrence sheets depend on the specification of the buffer in the call to `geo_ebird`. In the current example, we did not explore eBird records in a buffer around Piedmont NWR, but rather used only the actual refuge boundary.

Thus, the results include only seasonal abundance classifications, occurrence records, and summaries of eBird effort within the refuge boundaries:

``` r
make_checklists(piedmont)
```

![Output](./README-figs/geobird_screenshot.png)

If we specify an eBird query within the refuge boundaries as well as within a larger buffered area, we are gifted some additional occurrence information. If the buffer is larger than 5 km, we are also provided with some additional abundance information.

Take, for example, a comparison of eBird data within Piedmont NWR as well as a 10 km buffer around the refuge:

``` r
piedmont_buffs <- geo_ebird(SErefuges, which_polys = "Piedmont", buffers = c(0, 10))
make_checklists(piedmont_buffs)
```

Looking at the top panel below, provided enough complete checklists are available, we are now given (1) abundance codes in each of the distance categories (i.e., actual boundary vs. refuge + 10 km buffer), (2) the distance at which abundance estimates are first possible, (3) a "Status" that indicates whether an abundance classification was possible for at least one season within the refuge ("abundance-refuge") or only by adding the buffer around the refuge ("abundance-buffer"), and (4) a crude season-by-season comparison of relative abundance within the refuge boundary and within the larger area including the buffer.

For this last comparison, if the proportion of complete checklists containing a given species was &gt; 25% higher on the refuge compared to the larger, buffered area, it is marked as "on-refuge". This may indicate that the refuge houses more of a given species compared to the larger landscape (see, e.g., Pileated Woodpecker in winter). Conversely, an "off-refuge" designation indicates the prevalence of a given species (i.e., proportion of complete checklists) was higher on the larger landscape than on the refuge. A hyphen ("-") indicates they were more or less equal.

For the occurrence sheet (second panel from top), we are now given (1) the number of checklists reporting a given species at each distance category in each season, (2) the distance at which occurrence was first documented, and (3) a "Status" that indicates whether occurrence was documented within the refuge boundary ("occurrence-refuge") or only in the buffer ("occurrence-buffer").

The checklist (third panel) consolidates the abundance records (top panel) and occurrence only records (from the second panel) into an overall species list for the *actual* polygon boundary. Species that occur within the boundary but with inadequate data from an abundance classification are arbitrarily assigned a status of "Rare" for their season(s) of occurrence.

Effort is now also summarized for each distance category (bottom panel).

![Output with buffers](./README-figs/geobird_screenshot2.png)

### Visualizing records

We can also visualize the monthly patterns of eBird detections by species, relative to eBird effort (i.e., \# checklists), using the `plot_ebird_phen` function.

We can specify the species using the `species` argument (using the accepted AOU name but insensitive to capitalization) or, if not specified, we're presented with a list of available species form which to choose. We follow up on the possible discrepancy in Pileated Woodpecker abundance on-refuge and on the larger landscape.

``` r
plot_ebird_phen(piedmont_buffs)
```

![Pick a species, any species](./README-figs/PIWO.png)

![PIWO phenology](./README-figs/PIWO_plot.png)

The plot illustrates the higher proporortion of complete checklists containing Pileated Woodpeckers in fall and winter (Sep - Feb), and also provides information about the amount of monthly effort which, for Piedmont and the surrounding area, is greatest from February through June.

### Exporting to NWRSpecies Database

With version 0.3.3, specifying the `nwrspp = TRUE` argument to `make_checklist` generates an additional sheet in the output spreadsheet that is formatted to (hopefully) integrate seamlessly into the USFWS NWRSpecies database.

``` r
## Note, using the eBird query based only on refuge boundary
## No buffers
make_checklists(piedmont, nwrspp = TRUE)
```

![NWRSpecies tab](./README-figs/NWRSpecies.png)


NEWS.md
# geobird 0.3.3

* New option when outputting a spreadsheet to create an additional sheet in a format easy to incorporate into the NWRSpecies database.
* Fixed some bugs in how abundance and occurrence status were defined.

# geobird 0.3.2

* Substantial improvements to `plot_ebird_phen` to better handle plots including multiple sites (polygons) and allow the user to specify only a subset of site to be plotted
* Added conditions to avoid problems with polygons with adequate abundance data for all species and polygons without adequate data to generate a checklist

# geobird 0.3.1

* Add utility function of allow output of checklist to *.csv

# geobird 0.3

* Exclude hybrids and optional exclusion of "forms"
* Allow direct specification of file for parsing/storing in single call
* Rely more of data.table for improved performance (speed)
* General improvement of code clarity

# geobird 0.2

* Use data.table::fread to much faster reading of large text files
* Allow exclusion of polygons in addition to selection
* Improved output *.xls formatting
* Illustrate new features in README


R/make_checklists.R
#' Create spreadsheets of eBird species abundance and occurrence (i.e., # checklists).
#'
#' This function generates, for each queried polygon, a spreadsheet of all species detected on at
#'  least one checklist.  For each season (see Details), an estimated abundance classification
#'  is provided (see Details) if a minimum number of complete checklists \code{min_lists} is
#'  available.  Regardless, the total number of checklists (complete or not) on which each species
#'  was observed in each season is reported.  Likewise, the amount of effort (# of complete and
#'  total checklists) is reported by season.
#'
#' Seasons are classified as follows: spring (Mar - May), summer (Jun - Aug), fall (Sep - Nov),
#'  and winter (Dec - Feb).  Note, this corresponds roughly to the seasonal bird checklists
#'  present in many National Wildlife Refuge brochures (which may have had a small part in the
#'  decision to use this classification).  For monthly patterns of eBird detections by species,
#'  relative to eBird effort (i.e., # checklists), see \code{\link{plot_ebird_phen}}.
#'
#' Abundance designations are classified based on the proportion of complete checklists on
#'  which a given species occurs.  The \code{min_lists} argument allows the user to determine the
#'  number of checklists necessary to even attempt such a designation.  The default is 10 complete
#'  checklists, although a larger minimum (e.g., 50 or 100) is probably a more reasonable choice for
#'  generating an estimate of species relative abundance.  Abundance classifications (proportion of
#'  complete checklists) are:  Abundant (A; > 0.50), Common (C; (0.3 - 0.5]), Uncommon (U; (0.2 -
#'  0.3]), Occasional (O; (0.1 - 0.2]), Rare (R; (0.01 - 0.1]), and Vagrant (V; <= 0.01).  Notice
#'  that the cutoff for a "Vagrant" designation requires at least 100 checklists; with fewer checklists
#'  this designation is lumped with "Rare".
#'
#' @param geo_ebird_df a \code{\link{data.frame}} created by \code{\link{geo_ebird}}
#' @param min_lists integer or numeric vector of length = 1 indicating the minimum number of complete
#'  eBird checklists to require in a given season before assigning a seasonal abundance
#'  classification
#' @param exclude_form logical indicating whether observations from the slightly nebulous "form"
#'  category should be excluded (default = TRUE).  Setting to FALSE may add certain species (e.g.,
#'  Red Crossbill) but also more controversial entries (e.g., Northern Red-tailed Hawk)
#' @param xls logical indicating whether to output nicely formatted *.xls file (TRUE; default) or to
#'  create a \code{\link{list}} of relevant information without any formatting (FALSE; i.e., for
#'  further manipulation in R)
#' @param nwrspp logical indicating whether a sheet formatted for easy importing into the NWRSpecies
#'  database is requested (default = FALSE); only applies if `xls = TRUE`.
#' @param out_dir character string of the file path to the directory in which to save output
#'  spreadsheets
#' @return a Microsoft Excel spreadsheet for each distinct polygon (i.e., each (\code{name}) in
#'   \code{geo_ebird_df}) when \code{xls = TRUE} or a \code{\link{list}} of \code{\link{list}}s
#' @importFrom plyr "."
#' @export
#'
#' @examples
#' \dontrun{
#' # Get shapefile
#' SErefuges <- rgdal::readOGR("../GIS", "refuges", verbose = FALSE, stringsAsFactors = FALSE)
#'
#' # Query eBird records within Piedmont NWR and 10 km buffer
#' piedmont <- geo_ebird(SErefuges, which_polys = "Piedmont", buffers = c(0, 10))
#'
#' # Generate .xls spreadsheet contain abundance, occurrence, and effort information
#' # Require 100 complete checklists to generate abundance classification
#' make_checklists(piedmont, min_lists = 100)
#' }

make_checklists <- function(geo_ebird_df, min_lists = 10L, exclude_form = TRUE,
                            xls = TRUE, nwrspp = FALSE, out_dir = "../Output/")  {

    if (length(min_lists) != 1 | !(class(min_lists) %in% c("integer", "numeric")))
        stop("The min_lists argument must be an integer or numeric vector of length = 1")

    # Get number of buffers assessed
    n_buffs <- length(unique(geo_ebird_df$buff_dist_km))
    buff_dists <- sort(unique(geo_ebird_df$buff_dist_km))

    # Add season
    geo_ebird_df <- mutate(geo_ebird_df,
                           season = factor(get_season(lubridate::month(date)),
                                           levels = c("spring", "summer", "fall", "winter")))
    if (exclude_form) geo_ebird_df <- dplyr::filter(geo_ebird_df, category != "form")

    # Store taxonomic order info for later
    tax_order <- geo_ebird_df %>% dplyr::select(common_name, tax_order) %>%
        dplyr::group_by(common_name) %>% dplyr::summarise(tax_order = min(tax_order))

    # Tabulate the number of complete checklists by season (i.e., all_spp == 1)
    complete_checklists <- geo_ebird_df %>% dplyr::filter(all_spp == 1) %>%
        dplyr::group_by(name, season, buff_dist_km) %>%
        summarise(season_checklists = length(unique(checklist)))

    # Tabulate all checklists by season (complete or incomplete)
    all_checklists <- geo_ebird_df %>%
        group_by(name, season, buff_dist_km) %>%
        summarise(season_checklists = length(unique(checklist)))

    # If sufficient complete checklists, assign abundance classification by season
    abundance <- geo_ebird_df %>% filter(all_spp == 1) %>%
        left_join(complete_checklists, by = c("buff_dist_km", "name", "season")) %>%
        group_by(name, season, common_name, sci_name, buff_dist_km, season_checklists) %>%
        summarise(spp_checklists = n()) %>%
        mutate(p_lists = round(spp_checklists / season_checklists, 3),
               abundance = code_abundance(p_lists)) %>%
        mutate(abundance = ifelse(season_checklists >= min_lists, abundance, NA_character_)) %>%
        select(-season_checklists, -spp_checklists, -p_lists) %>%
        left_join(tax_order, by = "common_name") %>% data.table::as.data.table() %>%
        data.table::dcast(name + common_name + sci_name + tax_order ~ buff_dist_km + season,
                        value.var = "abundance", drop = c(TRUE, FALSE)) %>%
        as.data.frame()
    names(abundance)[-4:-1] <- flip_string(names(abundance)[-4:-1], add_on = "km")

    # Extract abundance at actual polygon boundaries only for use in final checklist generation
    if (0 %in% buff_dists) {
        no_abund_refuge <- apply(abundance[, 5:8], 1, function(row) all(is.na(row)))
        actual_abund <- abundance[!no_abund_refuge, 1:8]
    }

    # Tabulate total number of checklists recording a given species by season
    occurrence <- geo_ebird_df %>%
        left_join(all_checklists, by = c("buff_dist_km", "name", "season")) %>%
        group_by(name, season, common_name, sci_name, buff_dist_km, season_checklists) %>%
        summarise(n_checklists = n()) %>%
        select(-season_checklists) %>%
        left_join(tax_order, by = "common_name") %>% data.table::as.data.table() %>%
        data.table::dcast(name + common_name + sci_name + tax_order ~ buff_dist_km + season,
                          value.var = "n_checklists", drop = c(TRUE, FALSE)) %>%
        as.data.frame()
    names(occurrence)[-4:-1] <- flip_string(names(occurrence)[-4:-1], add_on = "km")

    # Extract occurrence at actual polygon boundaries only for use in final checklist generation
    if (0 %in% buff_dists) {
        no_occ_refuge <- apply(occurrence[, 5:8], 1, function(row) all(is.na(row)))
        actual_occ <- occurrence[!no_occ_refuge, 1:8]
    } else no_occ_refuge <- rep(TRUE, nrow(occurrence))

    # If comparing actual boundary and larger landscape (i.e., buffer >= 5 km),
    # document first detected distance and compare abundance classifications within boundary
    # and in larger landscape
    if (min(buff_dists) == 0 && max(buff_dists) >= 5) {

        abund_det_stats <- geo_ebird_df %>% filter(all_spp == 1) %>%
            select(name, common_name, buff_dist_km) %>% unique() %>%
            arrange(name, common_name, buff_dist_km)
        min_buff <- !duplicated(select(abund_det_stats, name, common_name))
        abund_det_stats <- abund_det_stats[min_buff, ]
        names(abund_det_stats)[3] <- "first_detected_km"
        abund_det_stats <- mutate(abund_det_stats,
                                  Status = ifelse(no_abund_refuge, "abundance-buffer",
                                                  "abundance-refuge"))

        abund_traj <- geo_ebird_df %>% filter(all_spp == 1) %>%
            left_join(complete_checklists, by = c("buff_dist_km", "name", "season")) %>%
            group_by(name, season, common_name, sci_name, buff_dist_km, season_checklists) %>%
            summarise(spp_checklists = n()) %>%
            filter(season_checklists >= min_lists) %>%
            mutate(p_lists = round(spp_checklists / season_checklists, 3)) %>%
            group_by(name, season, common_name) %>%
            arrange(buff_dist_km) %>%
            summarise(traj = ifelse(tail(p_lists, 1) - head(p_lists, 1) > 0.25, "off-refuge",
                                    ifelse (tail(p_lists, 1) - head(p_lists, 1) < -0.25, "on-refuge", "-"))) %>%
            reshape2::dcast(name + common_name ~ season, value.var = "traj")

        abundance <- left_join(abundance, abund_det_stats, by = c("name", "common_name")) %>%
            left_join(abund_traj, by = c("name", "common_name"))
    }

    # Document first detected distance; relevant with multiple buffers
    occ_det_stats <- geo_ebird_df %>%
        select(name, common_name, buff_dist_km) %>% unique() %>%
        arrange(name, common_name, buff_dist_km)
    min_buff <- !duplicated(select(occ_det_stats, name, common_name))
    occ_det_stats <- occ_det_stats[min_buff, ]
    names(occ_det_stats)[3] <- "first_detected_km"
    occ_det_stats <- mutate(occ_det_stats,
                            Status = ifelse(no_occ_refuge, "occurrence-buffer",
                                            "occurrence-refuge"))

    occurrence <- left_join(occurrence, occ_det_stats, by = c("name", "common_name"))

    # Summary table to complete checklists
    complete_checklists <- complete_checklists %>%
        reshape2::dcast(name + buff_dist_km ~ season, value.var = "season_checklists") %>%
        mutate(Effort = "Complete checklists only")

    # Summary table to all checklists
    all_checklists <- all_checklists %>%
        reshape2::dcast(name + buff_dist_km ~ season, value.var = "season_checklists") %>%
        mutate(Effort = "All checklists")

    # Consolidate effort into a single data frame
    checklists <- rbind(complete_checklists, all_checklists) %>% select(Effort, everything())
    # Change NAs to indicate to checklists
    checklists[, c("spring", "summer", "fall", "winter")][is.na(checklists[, c("spring", "summer", "fall", "winter")])] <- 0

    # Generate output spreadsheets, if requested
    if (xls) {
        poly_abund <- plyr::dlply(abundance, .(name), df_to_checklist,
                                  type = "abund", n_buffs, buff_dists)
        poly_occurrence <- plyr::dlply(occurrence, .(name), df_to_checklist,
                                       type = "occ", n_buffs, buff_dists)
        # Extract species with occurrence data only, combine with abundance data at actual boundary
        if (0 %in% buff_dists) {
            poly_final <- plyr::dlply(actual_abund, .(name), df_to_checklist,
                                      type = "final", n_buffs = 1, buff_dists = 0,
                                      actual_occ = actual_occ)
        }

        poly_effort <- plyr::dlply(checklists, .(name), select, -name)

        ## Create sheet for NWRSpecies incorporation
        if (nwrspp) {
            # Retrieve some internal tables for joining
            tlu <- system.file("extdata", "tlu_CMT.xlsx", package = "geobird") %>%
                xlsx::read.xlsx(1, stringsAsFactors = FALSE) %>%
                select(OrgName = ORGNAME, UnitCode = FBMS) %>%
                # Special case; rename Savannah
                mutate(OrgName = ifelse(OrgName == "Savannah-Pinckney National Wildlife Refuges",
                                     "Savannah National Wildlife Refuge", OrgName))
            taxon <- system.file("extdata", "MatchedBirds.xlsx", package = "geobird") %>%
                xlsx::read.xlsx(1, stringsAsFactors = FALSE) %>%
                select(CommonName, TaxonCode, Nativeness)

            max_buff <- max(buff_dists)
            old_abund_cols <- paste0(c("spring_", "summer_", "fall_", "winter_"), max_buff, "km")
            abund_cols <- paste0(c("Spring", "Summer", "Fall", "Winter"), "Abundance")
            keep_fields <- c(abund_cols, "Abundance", "Occurrence", "OccurrenceTag", "Category",
                             "CategoryCode", "UnitCode", "OrgName", "Nativeness", "CommonName",
                             "TaxonCode", "ExternalLinks")
            poly_nwrspp <- occurrence %>% dplyr::select_(.dots = c("name", "common_name", "Status")) %>%
                dplyr::left_join(., dplyr::select_(abundance, .dots = c("name", "common_name", old_abund_cols)),
                                 by = c("name", "common_name")) %>%
                mutate_(.dots = setNames(list(
                    ~ifelse(grepl("buffer", Status), "Unconfirmed", "Present"),
                    ~ifelse(grepl("buffer", Status), "Adjacent", NA_character_),
                    ~"Birds", # Add some constant columns
                    ~2,
                    ~"http://ebird.org"),
                    c("Occurrence", "OccurrenceTag", "Category", "CategoryCode", "ExternalLinks"))) %>%
                rename_(.dots = setNames(c(old_abund_cols, "name", "common_name"),
                                         c(abund_cols, "OrgName", "CommonName")))

            poly_nwrspp$Abundance <- poly_nwrspp[, abund_cols] %>% apply(., 1, compare) %>%
                ifelse(., poly_nwrspp[, abund_cols[1]], NA_character_)

            # Join ancillary information
            poly_nwrspp <- poly_nwrspp %>%
                insensitive(dplyr::left_join)(., tlu, by = "OrgName") %>%
                insensitive(dplyr::left_join)(., taxon, by = "CommonName") %>%
                # Final column arrangement
                dplyr::select_(.dots = keep_fields) %>%
                plyr::dlply(., .(OrgName))
        }

        for (name in names(poly_occurrence)) {

            if (substr(out_dir, nchar(out_dir), nchar(out_dir)) != "/") out_dir <- paste0(out_dir, "/")
            out_file <- paste0(out_dir, Cap(name), ".xls")

            # Do not output abundance sheet if no seasons meet the minimum checklist requirement
            # at the actual boundary (i.e., no abundance is generated beyond the actual boundary)
            if (!(0 %in% buff_dists) || all(is.na(poly_abund[[name]][, 2:5]))) {
                save.xlsx(poly_occurrence[[name]], out_file, sheetName = "eBird_all_records",
                          start.row = 3)
            } else {
                save.xlsx(poly_abund[[name]], out_file, sheetName = "eBird_abundance",
                          start.row = 3)
                save.xlsx(poly_occurrence[[name]], out_file, sheetName = "eBird_all_records",
                          start.row = 3, append = TRUE)
                if (name %in% names(poly_final))
                  save.xlsx(poly_final[[name]], out_file, sheetName = "Checklist",
                          start.row = 3, append = TRUE)
            }

            save.xlsx(poly_effort[[name]], out_file, sheetName = "eBird_effort", append = TRUE)

            if (nwrspp) save.xlsx(poly_nwrspp[[name]], out_file, sheetName = "NWRSpecies", append = TRUE)

        }

    } else {

        poly_abund <- arrange(abundance, tax_order) %>% plyr::dlply(.(name), select, -name)
        poly_occurrence <- arrange(occurrence, tax_order) %>%
            plyr::dlply(., .(name), select, -name)
        poly_effort <- plyr::dlply(checklists, .(name), select, -name)
        # Extract species with occurrence data only, combine with abundance data at actual boundary
        if (0 %in% buff_dists) {
            poly_final <- arrange(actual_abund, tax_order) %>%
                plyr::dlply(., .(name), function(df) {
                    poly_id <- unique(df$name)
                    poly_act_occ <- subset(actual_occ, name == poly_id)
                    spp_occ <- poly_act_occ$common_name[!(poly_act_occ$common_name %in% df$common_name)]
                    occ_only <- poly_act_occ[poly_act_occ$common_name %in% spp_occ, ]
                    # Convert to vagrancy status
                    occ_only[, 5:8] <- apply(occ_only[, 5:8], 2, as.character)
                    occ_only[, 5:8][!is.na(occ_only[, 5:8])] <- "V"
                    df <- rbind(df, occ_only) %>% select(-name)
                    df
                })

            out <- list(eBird_abundance = poly_abund,
                        eBird_all_records = poly_occurrence,
                        checklist = poly_final,
                        eBird_effort = poly_effort)
        } else {

            out <- list(eBird_abundance = poly_abund,
                        eBird_all_records = poly_occurrence,
                        eBird_effort = poly_effort)
        }

        out

    }

}
